File: poc/mobile/android/app/src/main/java/com/ironpal/poc/F16.kt
package com.ironpal.poc

import java.util.Base64

/**
 * IEEE 754 half-precision encode/decode for stored IMU windows (design §4.1, ledger Q11).
 * Layout: little-endian int16 halves, row-major [N][C]. Mirrored in src/model/f16.ts.
 */
object F16 {
  fun toHalf(f: Float): Int {
    val bits = java.lang.Float.floatToIntBits(f)
    val sign = (bits ushr 16) and 0x8000
    var value = (bits and 0x7fffffff) + 0x1000
    if (value >= 0x47800000) {
      if ((bits and 0x7fffffff) >= 0x47800000) {
        if (value < 0x7f800000) return sign or 0x7c00
        return sign or 0x7c00 or ((bits and 0x007fffff) ushr 13)
      }
      return sign or 0x7bff
    }
    if (value >= 0x38800000) return sign or ((value - 0x38000000) ushr 13)
    if (value < 0x33000000) return sign
    value = (bits and 0x7fffffff) ushr 23
    return sign or ((((bits and 0x7fffff) or 0x800000) + (0x800000 ushr (value - 102))) ushr (126 - value))
  }

  fun fromHalf(h: Int): Float {
    var mant = h and 0x03ff
    var exp = h and 0x7c00
    if (exp == 0x7c00) {
      exp = 0x3fc00
    } else if (exp != 0) {
      exp += 0x1c000
      if (mant == 0 && exp > 0x1c400) {
        return java.lang.Float.intBitsToFloat(((h and 0x8000) shl 16) or (exp shl 13) or 0x3ff)
      }
    } else if (mant != 0) {
      exp = 0x1c400
      do {
        mant = mant shl 1
        exp -= 0x400
      } while ((mant and 0x400) == 0)
      mant = mant and 0x3ff
    }
    return java.lang.Float.intBitsToFloat(((h and 0x8000) shl 16) or ((exp or mant) shl 13))
  }

  /** Encode [N][C] doubles → base64 of little-endian halves. */
  fun encode(series: Array<DoubleArray>): String {
    if (series.isEmpty()) return ""
    val c = series[0].size
    val out = ByteArray(series.size * c * 2)
    var k = 0
    for (row in series) for (j in 0 until c) {
      val h = toHalf(row[j].toFloat())
      out[k++] = (h and 0xff).toByte()
      out[k++] = ((h ushr 8) and 0xff).toByte()
    }
    return Base64.getEncoder().encodeToString(out)
  }

  /** Decode base64 halves → [N][channels]. */
  fun decode(b64: String, channels: Int): Array<DoubleArray> {
    if (b64.isEmpty() || channels <= 0) return emptyArray()
    val bytes = Base64.getDecoder().decode(b64)
    val n = bytes.size / (2 * channels)
    return Array(n) { i ->
      DoubleArray(channels) { j ->
        val k = (i * channels + j) * 2
        val h = (bytes[k].toInt() and 0xff) or ((bytes[k + 1].toInt() and 0xff) shl 8)
        fromHalf(h).toDouble()
      }
    }
  }
}

File: poc/mobile/src/model/f16.ts
// IEEE 754 half-precision windows, base64 of little-endian int16 halves, row-major [N][C].
// Mirrors android/.../F16.kt (design §4.1). Used by the inspector, tests and priors packaging.

import {Buffer} from 'buffer';

export function toHalf(val: number): number {
  const f32 = new Float32Array(1);
  const u32 = new Uint32Array(f32.buffer);
  f32[0] = val;
  const x = u32[0];
  const sign = (x >>> 16) & 0x8000;
  let m = (x >>> 12) & 0x07ff;
  const e = (x >>> 23) & 0xff;
  if (e < 103) {
    return sign;
  }
  if (e > 142) {
    return sign | 0x7c00 | (e === 255 && (x & 0x007fffff) ? 1 : 0);
  }
  if (e < 113) {
    m |= 0x0800;
    return sign | ((m >> (114 - e)) + ((m >> (113 - e)) & 1));
  }
  return sign | ((e - 112) << 10) | (m >> 1);
}

export function fromHalf(h: number): number {
  const s = (h & 0x8000) >> 15;
  const e = (h & 0x7c00) >> 10;
  const f = h & 0x03ff;
  if (e === 0) {
    return (s ? -1 : 1) * Math.pow(2, -14) * (f / Math.pow(2, 10));
  }
  if (e === 0x1f) {
    return f ? NaN : (s ? -1 : 1) * Infinity;
  }
  return (s ? -1 : 1) * Math.pow(2, e - 15) * (1 + f / Math.pow(2, 10));
}

export function encodeF16(series: number[][]): string {
  if (series.length === 0) {
    return '';
  }
  const c = series[0].length;
  const buf = Buffer.alloc(series.length * c * 2);
  let k = 0;
  for (const row of series) {
    for (let j = 0; j < c; j++) {
      buf.writeUInt16LE(toHalf(row[j]), k);
      k += 2;
    }
  }
  return buf.toString('base64');
}

export function decodeF16(b64: string, channels: number): number[][] {
  if (!b64 || channels <= 0) {
    return [];
  }
  const buf = Buffer.from(b64, 'base64');
  const n = Math.floor(buf.length / (2 * channels));
  const out: number[][] = new Array(n);
  for (let i = 0; i < n; i++) {
    const row = new Array<number>(channels);
    for (let j = 0; j < channels; j++) {
      row[j] = fromHalf(buf.readUInt16LE((i * channels + j) * 2));
    }
    out[i] = row;
  }
  return out;
}

/** CRC32 of the encoded window — the store's `checksum` (design §7.2). */
export function crc32(str: string): number {
  let c = ~0;
  for (let i = 0; i < str.length; i++) {
    c ^= str.charCodeAt(i) & 0xff;
    for (let k = 0; k < 8; k++) {
      c = (c >>> 1) ^ (0xedb88320 & -(c & 1));
    }
  }
  return ~c >>> 0;
}
